### PR TITLE
Ts 3612 - Restructuring client session and transaction processing

### PR DESCRIPTION
**Pull request by shinrich**
This has been running on a 5.3 base in our production in a few machines for about a month.  We are starting to roll it out more widely.  I've also run it in our production for a few days on a 6.1 base.  It had some performance issues, but the base 6.1 had the same performance issues which I've filed bugs on.  Also ran a short while with these changes on master.  Again some problems, but the same problems were on the base code in our environment as well.

These changes build upon James' abstraction of the ProxyClientSession.  We introduce a ProxyClientTransaction.  HttpSM works with the ProxyClientTransaction instead of the HttpClientSession.  Http1 and Http2 provide subclassed instances of the ProxyClientTransaction.  This takes FetchSM out of the picture for Http2.  https://www.dropbox.com/s/j87lph2z66vx05t/ProxyClientSessionRe-architectProposal.pdf?dl=0 contains the design document.  We did not do the Spdy aspects of the design.

SPDY is unchanged.  It still uses FetchSM and seems to work as well has it has been.  There are minimal changes in the Spdy code.

The bulk of the code changes are directly attributed to the Session/Transaction separation.  In addition, there are a few other changes.
* When moving to 6.1 I had memory freeing issues with some of the changes for SSL buffering introduced in TS-3714.  Since these changes were addressing issues in 5.0 that were fixed a different way in 5.2, I pulled out the extra buffering as unnecessary.  
* I made some changes in solve InkConInternal deletes (or lack of deletes) that occurred due to changes in the event_counts with the fetchSM removal.

### CODE REVIEW

**Comment by jpeach** [comment]
Please squash the clang-format commit. Can you elaborate on why this reverts TS-4321?

**Comment by shinrich** [comment]
I figured there would be some more updates via the review process so I delayed further squashes.  Will definitely squash before the final commit.

Since we aren't using FetchSM with this re-arrangement, it wasn't clear that we needed to do the function rearrangements and extra header copies and allocations.  Actually read the bug report more closely just now rather than just looking at the code changes.  It isn't clear to me that this is an issue without FetchSM, but I'll go head and apply the function changes here, and repush.

**Comment by maskit** [comment]
Reverting TS-4321 sounds reasonable if we can modify ``stream->response_header`` without any side effects.

The response headers in FetchSM must not be modified because FetchSM uses them internally to check if the response body is chunked. It seems ``Http2Stream::response_initialize_data_handling`` does the similar thing.

So all checks depend on response headers must be done before we modify headers here if we don't make the copy. If we cannot assure that, we still need the fix for TS-4321.

Moreover, generating H2 headers from headers in Http2Stream sounds odd.

File: proxy/ProxyClientSession.h [hunk 1]
```diff
@@ -25,14 +25,19 @@
...
 #include "InkAPIInternal.h"
+#include "http/HttpServerSession.h"
 
 // Emit a debug message conditional on whether this particular client session
 // has debugging enabled. This should only be called from within a client session
 // member function.
 #define DebugSsn(ssn, tag, ...) DebugSpecific((ssn)->debug(), tag, __VA_ARGS__)
 
+class ProxyClientTransaction;
+class AclRecord;
```
Comment (maskit) [inline proxy/ProxyClientSession.h]: I got this error.
```
In file included from ./Http1ClientSession.h:41:
../../proxy/ProxyClientSession.h:39:1: error: class 'AclRecord' was previously declared as a struct [-Werror,-Wmismatched-tags]
class AclRecord;
^
../../proxy/IPAllow.h:56:8: note: previous use is here
struct AclRecord {
       ^
../../proxy/ProxyClientSession.h:39:1: note: did you mean struct here?
class AclRecord;
^~~~~
struct
```

File: proxy/ProxyClientSession.h [hunk 2]
```diff
@@ -101,6 +106,75 @@ class ProxyClientSession : public VConnection
...
+    return false;
+  }
+
+  // Override if your session protocol cares
+  virtual void
+  set_half_close_flag(bool flag)
+  {
+  }
+  virtual bool
+  get_half_close_flag()
```
Comment (maskit) [inline proxy/ProxyClientSession.h]: This one should be const?
```
In file included from HttpSessionAccept.cc:26:
./Http1ClientSession.h:83:3: error: 'Http1ClientSession::get_half_close_flag' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
  get_half_close_flag() const
  ^
../../proxy/ProxyClientSession.h:124:3: note: hidden overloaded virtual function 'ProxyClientSession::get_half_close_flag' declared here: different qualifiers (none vs const)
  get_half_close_flag()
  ^
```

Comment (shinrich) [inline proxy/ProxyClientSession.h]: Yes, it should.  Fixing.  What compiler are you using?  It is giving better warnings.

Comment (maskit) [inline proxy/ProxyClientSession.h]: I'm using clang. That's why I use it.

**Comment by calavera** [comment]
@asfgit @jpeach I just want to leave a note here because we found an issue with this commit. Under certain circumstances ATS segfaults due to this changes. I haven't been able to find the real issue yet but it might be prudent to flag this commit before 7.0 ships. I'll open an issue as soon as I have more information.

**Comment by zwoop** [comment]
We should flag this for 6.2.0. Can you file a Jira please? @PSUdaemon

**Comment by shinrich** [comment]
@calavera could you share a stack trace?

**Comment by jpeach** [comment]
@calavera or crash logs :)

**Comment by calavera** [comment]
@jpeach, @shinrich I opened an issue:

https://issues.apache.org/jira/browse/TS-4444

**Comment by biilmann** [comment]
This pull request also seems to break one of our plugins that does HTTP requests in a background worker thread via `TSFetchUrl` (used indirectly through the `AsyncHttpFetch` C++ API).

The HTTP request will hang at this check for `the right thread` and never progress:

https://github.com/apache/trafficserver/pull/570/files#diff-40fc32ab3504372e0eb74a54108d1b6fR2404

Removing these checks for the right thread leads to a segfault.

**Pull request closed without merging**